cmd/goal/clerk.go

```diff
@@ -145,7 +146,13 @@ var sendCmd = &cobra.Command{
 		if txFilename == "" {
 			// Sign and broadcast the tx
 			wh, pw := ensureWalletHandleMaybePassword(dataDir, walletName, true)
-			tx, err := client.SendPaymentFromWallet(wh, pw, fromAddressResolved, toAddressResolved, fee, amount, noteBytes, closeToAddressResolved)
+			var tx transactions.Transaction
+
+			if firstValid != 0 && lastValid != 0 {
+				tx, err = client.SendPaymentFromWalletWithCustomValidity(wh, pw, fromAddressResolved, toAddressResolved, fee, amount, noteBytes, closeToAddressResolved, basics.Round(firstValid), basics.Round(lastValid))
```

Comment:
Since we only call this function with non-zero `firstValid` and `lastValid`, what do you think about adding those arguments to `SendPaymentFromWallet` and avoiding the two copies of functions?  Or is this for compatibility with potential third-party apps that are using `SendPaymentFromWallet`?

Reply:
Actually, we have a separate SDK repo that third-party apps should be using, so no backwards compat should be needed at this API level...

Reply:
@zeldovich Yes, you have it right: I tried to keep `ConstructPayment`/`SendPaymentFromWallet`'s interfaces and behavior the same. So, I implemented the new behavior as a new function, and tried to factor common behavior out where I could.

Reply:
Do we throw an error if someone specifies first and last round with range greater than 1000?

Reply:
@ejbaran No, goal (with this PR) allows the user to define any first and last round they like, even invalid ranges. The transaction wouldn't then make it into a block, as I believe it would fail validity checks, but goal lets the user create such a transaction. Should I add some protections?

Reply:
If it's low-risk to do so, yes. Most people will be using this to sign offline. If they are forced to specify a first and last they could easily make the error of specifying a large range and not know until they bring it back to submit.

Reply:
@ejbaran Added: if `firstvalid` and `lastvalid` are provided, they are checked against the current protocol `MaxTxnLifetime`, and an `error` occurs if that range is too large (or if lastValid is before firstValid)

Reply:
If there's no reason for the backwards compat, I'd prefer keeping a simpler API with a single function to do this, instead of introducing a second family of functions that take first/last valid.

Reply:
@zeldovich Gotcha. Pushing a new version now, is this more what you mean? I've nixed the extra functions, and instead changed `ConstructPayment`/`SendPaymentFromWallet` to take a first and last round - if you provide 0 and 0, they go to their old behavior.

N.B. this latest commit is broken and is meant mostly to show the fixed design (I'm currently running out the door), I haven't re-tested it and have not fixed the usages of `ConstructPayment` et al in other files to have the new extra `0,0` arguments. Can fix when I get back tonight :+1:

Reply:
Yeah, looking good, modulo some obvious fixes like lint, and removing the overly-strict check about `(lastValid == 0 || firstValid == 0) && ...`